src/Perl6/Metamodel/ContainerSpecProtocol.nqp

```diff
@@ -0,0 +1,34 @@
+role Perl6::Metamodel::ContainerSpecProtocol {
+    has $!code_pair;
+
+    method get_container_spec($obj) {
+        $!code_pair
+    }
+
+    method set_container_spec($obj, $code_pair) {
+        $!code_pair := $code_pair;
+    }
+
+    method publish_container_spec($obj) {
+        # If we have a container specification here, install that
+        if $!code_pair {
+            nqp::setcontspec($obj, 'code_pair', $!code_pair);
+            $obj.HOW.compose_repr($obj);
```

Comment:
Hm, I'm surprised to find this line here (and the similar one below); is it needed for something?

Reply:
That's for when the class.HOW itself had `set_container_spec` called on it.  Specifically for the `Proxy` class itself?

Reply:
Or are you implying that it should *always* do an nqp::setcontspec, even if `$!code_pair` hasn't been set?

Reply:
Aaah, I see.  The `.mro` has its own class as the first element, so it will find and handle the `container_spec` that way.

So addressed in fe1dcadaec

Reply:
Ah, I see. The `Proxy.HOW.compose_repr(Proxy);` line is specific to the bootstrap case, and so should have been left behind in `BOOTSTRAP`; it's not needed on the typical path.

Reply:
Addressed with 5eea5f4eb6